### PR TITLE
Geocode Trees Stretch Implementation

### PR DESCRIPTION
**Pull request by amartyanambiar**
Ideal Solution for #21 
This PR solves the problem of geocoding trees by using the Stretch-wise Implementation.
In the upcoming comments, you shall find detailed explanation for changes in the files.

This PR is to be merged into `geocode-trees`
After merge, within `geocode-trees`, work upon the generalisation of the test cases according to the UI selection shall be performed and incorporating validation steps too.

### Testing :
run `app.py`
run  `unpickle.py` with your desired nodes to check whether the shortest route is right.(path is correctly implemented)
Also check `graph.png`

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -68,8 +71,36 @@ def save():
...
+    gt.add_tree(tc.tree4.point_coordinate, start, end)
+    gt.add_tree(tc.tree5.point_coordinate, start, end)
+    gt.add_tree(tc.tree6.point_coordinate, start, end)
+    gt.save_and_return()
+
+
+test_stretch_1()
+test_stretch_2()
+##########################################
+
```
Comment (amartyanambiar) [inline app.py:101]: Test Case Functions for each stretch 
`basic_science to ganesha_statue` , `ganesha_statue to hostel_turn`

Utilised test_coordinates for test nodes.

File: cmritPath.py [hunk 2]
```diff
@@ -37,18 +37,19 @@ def __init__(self):
...
+
+        # adding edges
+        # G.add_edges_from([('teacher_parking', 'basic_science'),
+        #                   ('basic_science', 'teacher_parking'),
+        #                   ('basic_science', 'ganesha_statue'),
+        #                   ('ganesha_statue', 'basic_science'),
+        #                   ('ganesha_statue', 'hostel_turn'),
+        #                   ('hostel_turn', 'ganesha_statue'),
+        #                   ('volley_ball_court', 'hostel_turn'),
+        #                   ('hostel_turn', 'volley_ball_court')], length=300)
```
Comment (amartyanambiar) [inline cmritPath.py:49]: Commented this.
As we no longer are working by this approach.

Path shall be made btwn two nodes in a stretch only when trees are added.

File: geocodeTree.py [hunk 3]
```diff
@@ -10,17 +10,31 @@ class GeocodeTree:
...
+        self.tree_list = []
 
-    def add_tree(self, coord):
+    def add_tree(self, coord, start_node, end_node):
+        self.start_node = start_node
+        self.end_node = end_node
         self.tree_count += 1
         tree_name = f"t-{self.tree_count}"
-        nearest_node = ox.nearest_nodes(self.G, coord[0], coord[1])
+        self.tree_list.append(tree_name)
```
Comment (amartyanambiar) [inline geocodeTree.py:20]: a list is maintained to store the tree_node names

File: geocodeTree.py [hunk 4]
```diff
@@ -10,17 +10,31 @@ class GeocodeTree:
...
 
     def save_and_return(self):
+        for i in range(len(self.tree_list)-1):
+            self.G.add_edge(self.tree_list[i], self.tree_list[i+1])
+            self.G.add_edge(self.tree_list[i+1], self.tree_list[i])
+
+        self.G.add_edge(self.tree_list[0], self.start_node)
+        self.G.add_edge(self.start_node, self.tree_list[0])
+        self.G.add_edge(self.tree_list[-1], self.end_node)
+        self.G.add_edge(self.end_node, self.tree_list[-1])
```
Comment (amartyanambiar) [inline geocodeTree.py:35]: In this code snippet,

A for loop is run, to connect all consecutive tree nodes with each other bidirectionally.

Before : T1, T2, T3    (not connected)
After : T1 <=> T2 <=> T3  (connected bidirectionally)

Then the  first tree_node is connected to the start_node and the last tree_node is connected to the last_node of the stretch.

first_node <=> T1 <=> T2 <=> T3 <=> last_node

File: test_coordinates.py [hunk 5]
```diff
@@ -1,21 +1,25 @@
 from coordinates import Coordinates
 
 # basic_science
+basic_science = Coordinates("basic_science", [12.96626, 77.71211])
```
Comment (amartyanambiar) [inline test_coordinates.py:4]: added coordinates of nodes (basic_science, ganesha_statue, ganesha_statue, ganesha_statue)

**Review by Blake2912 — CHANGES_REQUESTED** [review]
```
['cmrit_entrance', 'teacher_parking', 'volley_ball_court', 'basic_science', 'ganesha_statue', 'hostel_turn', 't-1', 't-2', 't-3', 't-4', 't-5', 't-6']
    raise nx.NetworkXNoPath(f"No path between {source} and {target}.")
networkx.exception.NetworkXNoPath: No path between t-1 and volley_ball_court.
```
I m getting this error when I run the `unpickle.py` file @amartyanambiar can you look into what's the issue?

I m not sure if I am missing something I followed the above mentioned instructions

**Comment by Blake2912** [comment]
I haven't changed any node here I just ran it from the terminal that's it... so implicitly wouldn't an ideal case get fired up?

**Comment by amartyanambiar** [comment]
> I haven't changed any node here I just ran it from the terminal that's it... so implicitly wouldn't an ideal case get fired up?

Because I didn't push my unpickle.py.(futile push, nodes keep changing for test cases)
Change nodes in requested shortest_path statement.

**Pull request merged**